move: resolve function names for vm errors

rvantonder opened the pull request:
Fixes #7291

After chatting with @tnowacki, we go with the approach to first convert

```rs
impl From<VMError> for ExecutionError {
    fn from(error: VMError) -> Self { ... } ...}
``` 

to take a resolver to resolve function names. It is now:

```rs
pub fn convert_vm_error(
    error: VMError,
    fname_resolver: impl FunctionNameResolver, 
) -> ExecutionError { ... }

pub trait FunctionNameResolver {
    fn resolve(&self, m: &ModuleId, idx: FunctionDefinitionIndex) -> Option<Identifier>;                                                
}   
```

The resolver takes module ID and fn index (which we get from `error`) to resolve the function name. In `adapter.rs` the resolver uses `vm` state to resolve function names (what we want, now function names are resolving, yay) . 

In `id_leak_verifier.rs` we use an empty resolver (doesn't make a difference compared to previous logic). I'm not sure if/how to resolve the function name from this context. @tnowacki may have an idea, since he mentioned:

> the only spot that the function could be coming from is the current module we are checking, so we could give an invariant violation if that is nto true.
Really for (1) we should revert Sam’s change from last summer that forces the error type there to be VM error, and go back to having an associated type in the abstract interpreter

I feel like we can address that in a follow up PR if it's involved.

I'm not sure why the e2e test is failing, but it doesn't look specific to these changes (I see other open PRs failing on similar things?). I tried to run e2e tests locally without much success (against devnet and local network).

rvantonder commented on crates/sui-adapter/src/adapter.rs:
```diff
@@ -1353,7 +1390,31 @@ fn convert_type_argument_error(idx: usize, error: VMError) -> ExecutionError {
+}
+
+pub fn with_vm_resolver<
+    'r,
+    E: Debug,
+    S: ResourceResolver<Error = E> + ModuleResolver<Error = E>,
+>(
+    vm: &'r MoveVM,
+    state_view: &'r S,
+) -> impl Fn(&ModuleId, FunctionDefinitionIndex) -> Option<Identifier> + 'r {
```
![image](https://user-images.githubusercontent.com/888624/212436407-cd812a8e-1ec8-4fc6-9ea2-0261bfaeee6d.png)

Gonna be real, I just wanted to a helper function to create the resolver closure here. So re: lifetime annotations I just kind of tried things until the compiler was happy

amnn replied:
I would say Rust does not do so well with this functional style, I would suggest using traits:

```rust
trait FunctionNameResolver {
    fn resolve_fname(&self, m: &ModuleId, f: FunctionDefinitionIndex) ->  Option<Identifier>;
    fn convert_vm_error(&self, error: VMError) -> ExecutionError {
        /* ... */
    }
}

struct NullResolver;
impl FunctionNameResolver for NullResolver {
    fn resolve_fname(&self, _: &ModuleId, _: FunctionDefinitionIndex) -> Option<Identifier> {
        None
    }
}

struct VMResolver<'vm, E> {
    vm: &'vm MoveVM,
    state_view: &'vm impl MoveResolver,
}

impl FunctionNameResolver for VMResolver {
    fn resolve_fname(&self, m: &ModuleId, f: FunctionDefinitionIndex) -> Option<Identifier> {
        let module = vm.load_module(module_id, state_view).ok()?;
        let fdef = module.function_def_at(idx);
        let fhandle = module.function_handle_at(fdef.function);
        let fname = module.identifier_at(fhandle.name);
        Some(fname.to_owned())        
    }
}
```

Couple of other changes that I folded into this one:

- There's a convenient alias for the trait constraint on `state_view`: `MoveResolver`.
- I used `Result::ok` to convert it to an `Option` and then `?` syntax rather than `if let Ok(...) = ... { ...; Some(_) } else { None }`.

rvantonder replied:
This was super instructive, thank you! 

> There's a convenient alias for the trait constraint on state_view: MoveResolver.

Nice!

> I used Result::ok to convert it to an Option and then ? syntax rather than if let Ok(...) = ... { ...; Some(_) } else { None }

💯 

Some thoughts to tie up loose ends from my side:

- I played with the syntax a bit to get things to compile, ran into a couple things like not being able to do `impl MoveResolver` on the `state_view` value in the struct

- About `convert_vm_error`, I _think_ we want to keep this as a shared function in `error.rs` based on how things depend on it right now, and not a function that's part of the `FunctionNameResolver` trait. (It seems a bit strange to have something called `FunctionNameResolver` actually do error conversion and not just function resolving. Let me know if you had something else in mind here though).

amnn replied:
> About `convert_vm_error`, I think we want to keep this as a shared function in error.rs
 
I think this is fair!  No complaints from me on that.


> Not being able to do `impl MoveResolver` on the `state_view` in the struct.

This is either because the trait is not in scope (i.e. needs to be imported) or because Rust straight up doesn't support the impl syntax on structs.  I kind of suspect the latter?  In which case I *think* what you want to do here is to make the struct generic on `state_view`'s type and unconstrained, and then apply the constraint on your trait impl, something like:

```rust
struct VMResolver<'vm, S> {
    vm: &'vm MoveVM,
    state_view: &'vm S,
}

impl <'vm, S: MoveResolver> FunctionNameResolver for VMResolver<'vm, S> {
    /* ... */
} 
```

rvantonder replied:
Indeed, the latter! Your snippet is what I ended up arriving at 👍. Interestingly I think the compiler said it's an unstable/nightly feature.

amnn reviewed (approved):
Some notes on Rust idioms, but the change looks great, thanks for working on this @rvantonder !

rvantonder commented on crates/sui-types/src/messages.rs:
```diff
@@ -1615,11 +1616,16 @@ impl Display for MoveLocation {
-            "{module} in function definition {function} at offset {instruction}"
-        )
+        if let Some(fname) = function_name {
+            write!(f, "{module} in function {fname} at offset {instruction}")
```
Happy to do it, no strong feelings my side. @tnowacki, thoughts? I do see that some of the Move error strings use this format.

rvantonder commented on crates/sui-adapter/src/adapter.rs:
```diff
@@ -290,16 +303,28 @@ fn execute_internal<
+    let writes = match writes {
+        Ok(r) => r,
+        Err(e) => return Err(convert_vm_error(e, with_vm_resolver(vm, state_view))),
+    };
```
![](https://media.tenor.com/ySPZZZr5Vm0AAAAM/amazed-face.gif)

Very good, `map_err(..)?` was what I was missing!

tnowacki reviewed (changes requested):
I don't think the NullResolver is needed. We should always know the name of the function in the id leak case.

I'm also not a big fan of traits when a single function/lambda would do, as I think they make the code much more difficult to follow. Though I think we can safely kill the trait off if/when we make the absint change.

amnn commented:
I think if we get to a point where the null case doesn't need to be handled (i.e. there really is only one lambda) then I can get behind going back to a lambda, it's just when the lambda type is particularly hairy, or gets passed down multiple levels, that traits edge it out for me in terms of being documentable etc.  I agree this case is borderline though, and I don't feel particularly strong about it.

tnowacki commented:
> I think if we get to a point where the null case doesn't need to be handled (i.e. there really is only one lambda) then I can get behind going back to a lambda, it's just when the lambda type is particularly hairy, or gets passed down multiple levels, that traits edge it out for me in terms of being documentable etc. I agree this case is borderline though, and I don't feel particularly strong about it.

The null case isn't needed even for this PR. Sorry for not enough clarity on that.

In the ID leak verifier, there is only one case: The error comes from a function defined in the current module. So you can always just look up the name of the function in the current module, and if the module doesn't match, give an invariant violation instead (though that should be impossible... like really impossible) 

But to make that impossibility more clear, I'd be fine on waiting on this until we can land the update

rvantonder commented:
OK update on feedback, please take another look

> @tnowacki: I think we should maybe preserve the function index information? No need to hide the index? 

- Addressed error formatting. Around preserving index: I think it's useful for a compiler engineer when debugging, and less so for a user (adds noise IMO), but don't feel that strongly about this.

> @tnowacki: I don't think the NullResolver is needed [...] I'm also not a big fan of traits when a single function/lambda would do

- I removed need for a `NullResolver`. Because of that I also made the resolver a lambda again, and no longer the trait suggested by @amnn. 

> @tnowacki: Is this ever None now? (referring to whether we can resolve the function name to Some value)

- Unfortunately it's difficult to ensure it's never `None`, because the resolver function does `vm.load_module(module_id, ...);` which can fail. In order to ensure it's never `None`, we need to audit all callers will succeed loading a module, i.e., `vm.load_module` is always `Some`, and this would be a pretty heavy exercise. I could see restructuring this code into pipelines where we organize and ensure neatly when modules are resolvable, but that's a bigger ask. I think things are OK returning `Option<Identifier>` for now.

rvantonder commented on crates/sui-types/src/messages.rs:
```diff
@@ -1475,6 +1475,7 @@ pub struct MoveLocation {
     pub module: ModuleId,
     pub function: u16,
     pub instruction: CodeOffset,
+    pub function_name: Option<String>,
```
See comment: https://github.com/MystenLabs/sui/pull/7306#issuecomment-1404093925. The crux is that this `vm.load_module` may fail in the resolver, making the value optional: https://github.com/MystenLabs/sui/pull/7306/files#diff-093491827b897090f930c1656a3c6518a1212ae5ddf6d9786eb85d272c0bb557R1394

It's not impossible to overcome, but looks like tricky. I think we can roll with `Option` and revisit separately.

amnn commented on crates/sui-adapter/src/adapter.rs:
```diff
@@ -1345,15 +1365,36 @@ fn missing_unwrapped_msg(id: &ObjectID) -> String {
     ExecutionErrorKind::entry_type_argument_error(idx as TypeParameterIndex, kind).into()
 }
+
+pub fn with_vm_resolver<
```
Is it fair to say that the only kind of lambda that gets passed to `convert_vm_error` is one that comes from calling this function, and if so, does it make sense for `convert_vm_error` to just accept the `vm` and `state_view` directly, and, in effect inline the call to the lambda?

rvantonder replied:
I thought about this too. I think the trouble comes when we want to use `convert_vm_error` when we _don't_ have a handle on `vm` or `state_view` (in which case, it's handy to have a resolver function to do things). 

It might be that the intent behind `convert_vm_error` is to say "hey yeah we should always have a valid handle to vm for these errors. _However_ prior to this PR, there definitely was a case where we didn't have a handle on `vm`, and used `convert_vm_error`, so the intent here was maybe more aspirational than anything else. So yes, _now_ that I've shifted things around, it's become obvious that it's possible to invoke `convert_vm_error` with a valid `vm` always. 

So: Now I'm OK with passing `vm` and `state_view` directly. I will wait for @tnowacki to weigh in on this, versus passing a lambda.

rvantonder replied:
More details on how we resolved errors and didn't have a handle on `vm`: https://github.com/MystenLabs/sui/pull/7306/files#r1087087209

Essentially, the `convert_vm_error` function is a bit _too_ much of a catch-all error resolution function, so teasing a part the actual state/errors possible at various points during execution is going to be the thing that unlocks better errors. Similar story with the location info issue in `id_leak_verifier.rs` really.

tnowacki replied:
But we could just shuffle around publish so we have the VM instance? there isn't really a reason not to

rvantonder replied:
works for me, done!

rvantonder commented on crates/sui-adapter/src/adapter.rs:
```diff
@@ -337,7 +347,13 @@ pub fn publish<
         })
-        .collect::<move_binary_format::errors::VMResult<Vec<CompiledModule>>>()?;
+        .collect::<move_binary_format::errors::VMResult<Vec<CompiledModule>>>()
+        .map_err(|e| {
+            ExecutionError::new_with_source(
+                sui_types::messages::ExecutionFailureStatus::VMInvariantViolation,
+                e,
+            )
+        })?;
```
This change is what's needed to get rid of the previous `NullResolver`: in order to invoke `convert_vm_error`, we need a handle on `vm`. But in this `publish` function, there is no `vm` handle (and no need for it based on the runtime).

I dug into how this could possibly fail and what error it produces. After looking at `CompiledModule::deserialize(b)` and  `e.finish`, it was clear that this can only fail with [`VERIFIER_INVARIANT_VIOLATION`](https://sourcegraph.com/github.com/move-language/move/-/blob/language/move-binary-format/src/deserializer.rs?L54-56&subtree=true), code `2016` which ends up being converted to `sui_types::messages::ExecutionFailureStatus::VMInvariantViolation` via:

https://sourcegraph.com/github.com/move-language/move/-/blob/language/move-core/types/src/vm_status.rs?L705-709&subtree=true

https://sourcegraph.com/github.com/move-language/move/-/blob/language/move-core/types/src/vm_status.rs?L29-32&subtree=true

https://sourcegraph.com/github.com/MystenLabs/sui/-/blob/crates/sui-types/src/error.rs?L759&subtree=true

tnowacki replied:
Sorry, I'm not sure that I follow. The error here could be a result from any invalid series of bytes attempting to be published. So why are we mapping the error to VM Invariant violation? 

e.g. if I pass in `publish(.., vec![vec![0]], ..)` I should get a deserialization error. Won't this code here just switch that into an invariant violation error?

rvantonder replied:
...you're right. I overlooked this somehow and this is dumb.

> [...] shuffle around publish so we have the VM instance?

after passing `vm` in `publish`, we can just call out to `convert_vm_error` here, so it's fixed

tnowacki commented on crates/sui-adapter/src/adapter.rs:
```diff
@@ -1345,15 +1365,36 @@ fn missing_unwrapped_msg(id: &ObjectID) -> String {
+    state_view: &'r S,
+) -> impl Fn(&ModuleId, FunctionDefinitionIndex) -> Option<Identifier> + 'r {
+    |m: &ModuleId, idx: FunctionDefinitionIndex| {
+        let module = vm.load_module(m, state_view).ok()?;
```
I am wondering now... is this safe? I'm not sure that it is. I don't think we have a test for it, but could you try publishing a module that has a Move level verifier error?

There won't be a module to load in that case. I am logging off here shortly, but can help craft such an example tomorrow if you need the help :)

rvantonder replied:
Not sure I follow what could be unsafe here, are you referring to this line and the call to `ok()`?

```rust
let module = vm.load_module(m, state_view).ok()?;
```

Because I believe `ok()` just [converts `Result` to `Option`](https://sourcegraph.com/github.com/rust-lang/rust/-/blob/library/core/src/result.rs?L642-652&subtree=true) (don't think this is unsafe?)

I can look into a test, but it'd be great to understand what looks unsafe here.

tnowacki replied:
>  it'd be great to understand what looks unsafe here.

Unsafe here meaning it won't do what you want :) not like, panic or anything. 

Let's say you publish this module 
```
// example.mvir
module 0x0.m {

fun no(): u64 {
  0u8
}

}
```
This will go through and eventually in the adapters `verify_and_link` it will call through the VM to `session.publish_module_bundle` which eventually calls the bytecode verifier. That will fail, and produce a `VMError` which points to `0x0::m::no` at offset `0`. We will then try to convert the error, and then load the module. But that module won't exist! So instead of getting a function name of `Some("no")` you will get `None`

rvantonder replied:
Got it, so we are possibly missing a case where we could report module and function but don’t (false negative) for that code path. 

Proposal though: we weren’t doing this before anyway right (reporting functions for this case you're describing), so this isn’t making the state of things worse, and with these changes we do fix up a good number of MoveAbort issues. How about I file an issue for this case, assign it, and investigate, and handle it properly separately after this bigger change? I also want to bring back module/function location reporting for the id_leak_verifier too.

tnowacki replied:
> How about I file an issue for this case, assign it, and investigate, and handle it properly separately after this bigger change?

Sure, I can take a look after this lands :)

rvantonder commented:
Addressed more comments, and made the call to pass `vm` and `state_view` to `convert_vm_error` directly and inline, instead of lambdas/closures. Think we're close now, please take another look @tnowacki 🙏

tnowacki reviewed (approved):
Thanks for all the iterations on this! And sorry for being so persnickety 

Hopefully we can keep improving this and get some more location information in other places!

rvantonder commented:
no worries, I don't want things to degrade either, appreciate the diligence!
